src/main/java/com/onlinebookstore/onlinebookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+import com.onlinebookstore.onlinebookstore.exeption.EntityNotFoundException;
+import com.onlinebookstore.onlinebookstore.mapper.CartItemMapper;
+import com.onlinebookstore.onlinebookstore.mapper.ShoppingCartMapper;
+import com.onlinebookstore.onlinebookstore.model.Book;
+import com.onlinebookstore.onlinebookstore.model.CartItem;
+import com.onlinebookstore.onlinebookstore.model.ShoppingCart;
+import com.onlinebookstore.onlinebookstore.model.User;
+import com.onlinebookstore.onlinebookstore.repository.BookRepository;
+import com.onlinebookstore.onlinebookstore.repository.CartItemRepository;
+import com.onlinebookstore.onlinebookstore.repository.ShoppingCartRepository;
+import com.onlinebookstore.onlinebookstore.repository.UserRepository;
+import com.onlinebookstore.onlinebookstore.service.interfaces.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private static final Logger logger = LoggerFactory.getLogger(ShoppingCartServiceImpl.class);
+
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public ShoppingCartResponseDto getCartByUser(Long userId) {
+        logger.info("Fetching cart for user with id {}", userId);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseGet(() -> createShoppingCartForUser(userId));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto addBookToCart(Long userId,
+                                                 CartItemRequestDto cartItemRequestDto) {
+        logger.info("Adding book to cart for user with id {}", userId);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseGet(() -> createShoppingCartForUser(userId));
+
+        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(() -> new RuntimeException("Book not found"));
+
+        CartItem cartItem = cartItemMapper.toModel(cartItemRequestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItem.setBook(book);
+        cartItem = cartItemRepository.save(cartItem);
+
+        shoppingCart.getCartItems().add(cartItem);
+        shoppingCart = shoppingCartRepository.save(shoppingCart);
+
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto updateCartItem(Long cartItemId, int quantity) {

```

Comment:
can we also validate in that method that this item actually belongs to the user who has sent that request? because otherwise it would not be very secure